### DISCUSSION TITLE
Why is the saving of temporary files handled this way?

### DISCUSSION THREAD

**Discussion by pausto** [open-meteo/open-meteo#689] [Q&A] [answered]
Hello
I have no experience in swift programming. I am researching the code and I was interested in this point: storing temporary files on download. The file is created by fn and immediately deleted from the file system.  A few questions arose:
Where can I read about this technique?
How does this work?
Why is it done this way? 
Wouldn't there be problems with the file system when writing files this way?

Upd. When download multiple files, the folder size and free space does not change. At the same time all files are saved and accessed via handle.

Thanks

P.s. there is a double slash in the path.

![om_tmp_files](https://github.com/open-meteo/open-meteo/assets/3185083/bf1ec2d0-5aaf-41ce-9e09-2ad1d2aeff7e)

**Comment by patrick-zippenfenig** [comment, marked as answer]
Hi, this behaviour is not specific to Swift, but how posix file systems work. Deleting a "file" only unreferences (unlink) its path, but if the file is currently opened by a process, the file remains. Once the process releases the file (or the process terminates) the file is finally deleted from the file system. This is very helpful to store data temporarily that does not fit in memory, while automatically deleting all temporary files in case the process crashes.

The reason why I am doing this is: The original weather data is stored as a 2 dimensional map for each time step. Open-Meteo however, converts data into a time-series to quickly access it. Therefore data needs to be transposed. Although transposing data is a relatively a simple operation to switch dimensions from `[x, y, time]` to `[time, x, y]`, it gets more complicated, once the amount of data does not fit into memory anymore. IIRC generating a yearly file for ERA5-Land would require round 250 GB memory.

To divide the problem into smaller chunks, the download process takes each individual timestep into smaller cubes of data ([around 5000 locations](https://github.com/open-meteo/open-meteo/blob/main/Sources/App/Gfs/GfsDownload.swift#L219)). While downloading, each time-step is stored into temporary file. Depending on the weather model up to 60 GB of temporary files are generated. After all time-steps have been downloaded to temporary files the conversion to time-series starts, loops over all temporary files and updates the underlaying database (5000 locations a time). Inside the Open-Meteo source code, I refer to it as `streaming`. 

Due to the way the linux kernel caches files in its page cache and if the server has more system memory than the working set, all data might be still in memory and the no disk-reads are required. This can improve data-conversion speed. Note: Currently, data is flushed to disk for temporary files, but this could be skipped, so the actual data is never written to disk in the first place. It would only get written to disk if insufficient system memory is available (see page cache dirty ratio).

Until recently, those temporary files remained on the file-system and were not immediately deleted. E.g. `./data/download-gfs025/temperature_2m_timestep100.om`. This increasingly caused issues, because delays in model updates may arrive so late, that the file `./data/download-gfs025/temperature_2m_timestep100.om` got overwritten by the next weather model run. Most of the time, the download crashed, but in rare occurrences it generated faulty data. By using temporary files, this cannot happen anymore.

The lastest round of optimisations have been done in order to process the CMA GRAPES model. A single run contains 120 GB of GRIB files. Just converting the GRIB file with regular GRIB tools takes multiple hours. As a weather forecast is best "served fresh", I upgraded downloading, decoding and data conversion to fully [concurrent code](https://github.com/open-meteo/open-meteo/blob/main/Sources/App/CMA/CmaDownloader.swift#L213). Swift is actually very helpful in this regard with `async sequences` and automatic code checking using `Sendable`. 

The Open-Meteo code might not be the best reference to study Swift code. To manage the amount of weather data efficiently, I highly optimised code to less obvious cases. Most of those optimisations were necessary to keep memory utilisation at reasonable levels and process large amount of weather data quickly. 

Let me know if you have any more questions!
